onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -730,60 +754,41 @@ common::Status SparseTensorProtoToDenseTensorProto(const ONNX_NAMESPACE::SparseT
   const auto& indices = sparse.indices();
   auto dims = gsl::make_span<const int64_t>(dense.dims().data(), dense.dims().size());
 
-  // need to read in sparse data first as it could be in a type specific field, in raw data, or in external data
-  size_t sparse_bytes = 0;
-  ORT_RETURN_IF_ERROR(GetSizeInBytesFromTensorProto<0>(sparse_values, &sparse_bytes));
-
   if (type != TensorProto_DataType_STRING) {
-    std::vector<unsigned char> sparse_data_storage(sparse_bytes, 0);
-    void* sparse_data = sparse_data_storage.data();
-
+    // need to read in sparse data first as it could be in a type specific field, in raw data, or in external data
+    size_t sparse_bytes = 0;
+    std::unique_ptr<uint8_t[]> sparse_data_storage;
+    ORT_RETURN_IF_ERROR(UnpackInitializerData(sparse_values, sparse_data_storage, sparse_bytes));
+    void* sparse_data = sparse_data_storage.get();
     size_t element_size = 0;
-
-    // setup buffer for output
-    switch (type) {
-      case TensorProto_DataType_FLOAT: {
-        element_size = sizeof(float);
-        UnpackTensor<float>(sparse_values, static_cast<float*>(sparse_data), n_sparse_elements);
-        break;
-      }
-      case TensorProto_DataType_INT64: {
-        element_size = sizeof(int64_t);
-        UnpackTensor<int64_t>(sparse_values, static_cast<int64_t*>(sparse_data), n_sparse_elements);
-        break;
-      }
-      case TensorProto_DataType_INT32: {
-        element_size = sizeof(int32_t);
-        UnpackTensor<int32_t>(sparse_values, static_cast<int32_t*>(sparse_data), n_sparse_elements);
-        break;
-      }
-      case TensorProto_DataType_DOUBLE: {
-        element_size = sizeof(double);
-        UnpackTensor<double>(sparse_values, static_cast<double*>(sparse_data), n_sparse_elements);
-        break;
-      }
-      case TensorProto_DataType_UINT32: {
-        element_size = sizeof(uint32_t);
-        UnpackTensor<uint32_t>(sparse_values, static_cast<uint32_t*>(sparse_data), n_sparse_elements);
-        break;
-      }
-      case TensorProto_DataType_UINT64: {
-        element_size = sizeof(uint64_t);
-        UnpackTensor<uint64_t>(sparse_values, static_cast<uint64_t*>(sparse_data), n_sparse_elements);
-        break;
-      }
-      default:
-        status = ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Unsupported sparse tensor data type of ", type);
-    }
+#if !defined(ORT_MINIMAL_BUILD)
+    MLTypeCallDispatcherRet<Status, GetElementSize, float, int64_t, int32_t, double, uint32_t, uint64_t, BFloat16> type_disp(type);
+#else
```

Comment:
Do we need to support BFloat16 in a full build? Doesn't look like the existing code did so wondering if we need to add that support in this PR or not.

Reply:
Also wondering if we need to support int8_t/uint8_t given 8-bit quantization seems to be used more frequently these days

---
In reply to: [510541616](https://github.com/microsoft/onnxruntime/pull/5540#discussion_r510541616) [](ancestors = 510541616)